File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Fab.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * KinetixFab — mirrors `packages/ui/src/components/fab.tsx` (`fabVariants`).
 * `size-14`/`size-11` (56dp/44dp) aren't on the shared `spacing_*` scale —
 * hardcoded, same reasoning as `KinetixToggle`'s off-scale sizes. `shadow-lg`
 * has no elevation token in this package (see `KinetixCard`'s doc comment)
 * — approximated with a larger literal than Card's `shadow-sm` stand-in.
 * Hover/active color shifts aren't ported — same "no hover/focus states"
 * gap as `KinetixButton`.
 */
enum class KinetixFabVariant { Primary, Secondary }
enum class KinetixFabSize { Default, Sm }

@Composable
fun KinetixFab(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: KinetixFabVariant = KinetixFabVariant.Primary,
    size: KinetixFabSize = KinetixFabSize.Default,
    extended: Boolean = false,
    enabled: Boolean = true,
    content: @Composable () -> Unit,
) {
    val colors = KinetixColorScheme.current
    val diameter: Dp = when (size) {
        KinetixFabSize.Default -> 56.dp // size-14, not on the shared scale
        KinetixFabSize.Sm -> 44.dp // size-11, not on the shared scale
    }
    val containerColor = when (variant) {
        KinetixFabVariant.Primary -> colors.primary
        KinetixFabVariant.Secondary -> colors.secondary
    }
    val contentColor = when (variant) {
        KinetixFabVariant.Primary -> colors.primaryForeground
        KinetixFabVariant.Secondary -> colors.secondaryForeground
    }
    val alpha = if (enabled) 1f else 0.5f
    val shape = CircleShape

    var fabModifier = modifier.height(diameter)
    fabModifier = if (extended) fabModifier.widthIn(min = diameter) else fabModifier.width(diameter)
    fabModifier = fabModifier
        .shadow(6.dp, shape)
        .clip(shape)
        .background(containerColor.copy(alpha = alpha))
        .clickable(enabled = enabled, onClick = onClick)
        .padding(if (extended) PaddingValues(horizontal = dimensionResource(R.dimen.spacing_5)) else PaddingValues(0.dp))

    Row(
        modifier = fabModifier,
        horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_2), Alignment.CenterHorizontally),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CompositionLocalProvider(LocalContentColor provides contentColor.copy(alpha = alpha)) {
            content()
        }
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/Fab.swift
//
// Fab.swift — KinetixFab.
//
// Mirrors packages/ui/src/components/fab.tsx (`fabVariants`): a floating
// action button, circular by default or an extended pill. `size-14` /
// `size-11` (56 / 44) are off the shared spacing scale. `shadow-lg` has
// no elevation token — a larger literal than KinetixCard's stand-in.
// Hover/active colour shifts aren't ported (no hover on touch) — same
// gap as KinetixButton.
//

import SwiftUI

public enum KinetixFabVariant {
    case primary, secondary
}

public enum KinetixFabSize {
    case `default`, sm
}

public struct KinetixFab<Content: View>: View {
    @Environment(\.kinetixColors) private var colors
    @Environment(\.isEnabled) private var isEnabled

    private let variant: KinetixFabVariant
    private let size: KinetixFabSize
    private let extended: Bool
    private let action: () -> Void
    private let content: Content

    public init(
        variant: KinetixFabVariant = .primary,
        size: KinetixFabSize = .default,
        extended: Bool = false,
        action: @escaping () -> Void,
        @ViewBuilder content: () -> Content
    ) {
        self.variant = variant
        self.size = size
        self.extended = extended
        self.action = action
        self.content = content()
    }

    private var diameter: CGFloat { size == .default ? 56 : 44 }

    private var background: Color {
        variant == .primary ? colors.primary : colors.secondary
    }

    private var foreground: Color {
        variant == .primary ? colors.primaryForeground : colors.secondaryForeground
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: 8) { content }
                .font(.system(size: size == .default ? 24 : 20, weight: .medium))
                .foregroundStyle(foreground)
                .frame(width: extended ? nil : diameter, height: diameter)
                .frame(minWidth: extended ? diameter : nil)
                .padding(.horizontal, extended ? 20 : 0) // px-5
                .background(background, in: Capsule())
                .shadow(color: .black.opacity(0.2), radius: 8, y: 4) // shadow-lg approx
        }
        .buttonStyle(.plain)
        .opacity(isEnabled ? 1 : 0.5)
    }
}
